pkg/services/authz/rbac/store/permission_store.go

```diff
@@ -0,0 +1,159 @@
+}
+
+type PermissionsQuery struct {
+	OrgID         int64
+	UserID        int64
+	Action        string
+	ActionSets    []string
+	TeamIDs       []int64
+	Role          string
+	IsServerAdmin bool
+}
+
+func NewSQLPermissionStore(sql legacysql.LegacyDatabaseProvider, tracer tracing.Tracer) *SQLPermissionsStore {
+	return &SQLPermissionsStore{sql, tracer}
+}
+
+var _ PermissionStore = (*SQLPermissionsStore)(nil)
+
+type SQLPermissionsStore struct {
+	sql    legacysql.LegacyDatabaseProvider
+	tracer tracing.Tracer
+}
+
+var sqlUserPerms = mustTemplate("permission_query.sql")
+
+type getPermissionsQuery struct {
+	sqltemplate.SQLTemplate
+	Query *PermissionsQuery
+
+	PermissionTable  string
+	UserRoleTable    string
+	TeamRoleTable    string
+	BuiltinRoleTable string
+}
+
+func (r getPermissionsQuery) Validate() error {
+	return nil
+}
+
+func newGetPermissions(sql *legacysql.LegacyDatabaseHelper, q *PermissionsQuery) getPermissionsQuery {
+	return getPermissionsQuery{
+		SQLTemplate:      sqltemplate.New(sql.DialectForDriver()),
+		Query:            q,
+		PermissionTable:  sql.Table("permission"),
+		UserRoleTable:    sql.Table("user_role"),
+		TeamRoleTable:    sql.Table("team_role"),
+		BuiltinRoleTable: sql.Table("builtin_role"),
+	}
+}
+
+func (s *SQLPermissionsStore) GetUserPermissions(ctx context.Context, ns claims.NamespaceInfo, query PermissionsQuery) ([]accesscontrol.Permission, error) {
+	ctx, span := s.tracer.Start(ctx, "authz_direct_db.database.GetUserPermissions")
+	defer span.End()
+
+	sql, err := s.sql(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	query.OrgID = ns.OrgID
+	req := newGetPermissions(sql, &query)
+	q, err := sqltemplate.Execute(sqlUserPerms, req)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := sql.DB.GetSqlxSession().Query(ctx, q, req.GetArgs()...)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if res != nil {
+			_ = res.Close()
+		}
+	}()
+
+	var perms []accesscontrol.Permission
+	for res.Next() {
+		var perm accesscontrol.Permission
+		if err := res.Scan(&perm.Kind, &perm.Attribute, &perm.Identifier, &perm.Scope); err != nil {
+			return nil, err
+		}
+		perms = append(perms, perm)
+	}
+
+	return perms, nil
+}
+
+var _ PermissionStore = (*StaticPermissionStore)(nil)
+
+func NewStaticPermissionStore(ac accesscontrol.Service) *StaticPermissionStore {
+	return &StaticPermissionStore{ac}
+}
+
+type StaticPermissionStore struct {
+	ac accesscontrol.Service
+}
+
+func (s *StaticPermissionStore) GetUserPermissions(ctx context.Context, ns claims.NamespaceInfo, query PermissionsQuery) ([]accesscontrol.Permission, error) {
+	roles := []string{query.Role}
+	if query.IsServerAdmin {
+		roles = append(roles, "Grafana Admin")
+	}
+
+	static := s.ac.GetStaticRoles(ctx)
+
+	var permissions []accesscontrol.Permission
+	for _, name := range roles {
+		r, ok := static[name]
+		if !ok {
+			continue
```

Comment:
nit: let's add a debug log here, it should never happen.

Reply:
Ah, I see, for enterprise we will not have any roles, so we'll always be hitting this condition. Makes sense to not have a log then.